go/vt/tabletserver/queryservice/wrapped.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2017, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package queryservice
+
+import (
+	"golang.org/x/net/context"
+
+	"github.com/youtube/vitess/go/sqltypes"
+	"github.com/youtube/vitess/go/vt/tabletserver/querytypes"
+
+	querypb "github.com/youtube/vitess/go/vt/proto/query"
+)
+
+var _ QueryService = &wrappedService{}
+
+// WrapperFunc defines the signature for the wrapper function used by Wrap.
+// Parameter ordering is as follows: original parameters, connection, method name, additional parameters and inner func.
+type WrapperFunc func(ctx context.Context, target *querypb.Target, conn QueryService, name string, inTransaction, isStreaming bool, inner func(context.Context, *querypb.Target, QueryService) error) error
+
+// Wrap returns a wrapped version of the original QueryService implementation.
```

Comment:
This comment would be much easier to read if you mention an example usage of this very early on.

I think you should mention the discoverygateway as user and why this is helpful (no need to repeat the code).